src/fpm/manifest/preprocess.f90

```diff
@@ -0,0 +1,192 @@
+
+      !> Macros to be defined for the preprocessor
+      type(string_t), allocatable :: macros(:)
+
+   contains
+
+      !> Print information on this instance
+      procedure :: info
+
+   end type preprocess_config_t
+
+contains
+
+   !> Construct a new preprocess configuration from TOML data structure
+   subroutine new_preprocess_config(self, table, error)
+
+      !> Instance of the preprocess configuration
+      type(preprocess_config_t), intent(out) :: self
+
+      !> Instance of the TOML data structure.
+      type(toml_table), intent(inout) :: table
+
+      !> Error handling
+      type(error_t), allocatable, intent(inout) :: error
+
+      call check(table, error)
+      if (allocated(error)) return
+
+      call table%get_key(self%name)
+
+      call get_list(table, "suffixes", self%suffixes, error)
+      if (allocated(error)) return
+
+      call get_list(table, "directories", self%directories, error)
+      if (allocated(error)) return
+
+      call get_list(table, "macros", self%macros, error)
+      if (allocated(error)) return
+
+   end subroutine new_preprocess_config
+
+   !> Check local schema for allowed entries
+   subroutine check(table, error)
+
+      !> Instance of the TOML data structure.
+      type(toml_table), intent(inout) :: table
+
+      !> Error handling
+      type(error_t), allocatable, intent(inout) :: error
+
+      character(len=:), allocatable :: name
+      type(toml_key), allocatable :: list(:)
+      logical :: suffixes_present, directories_present, macros_present
+      integer :: ikey
+
+      suffixes_present = .false.
+      directories_present = .false.
+      macros_present = .false.
+
+      call table%get_key(name)
+      call table%get_keys(list)
+
+      do ikey = 1, size(list)
+         select case(list(ikey)%key)
+          case default
+            call syntax_error(error, "Key " // list(ikey)%key // "is not allowed in preprocessor"//name)
+            exit
+          case("suffixes")
+            suffixes_present = .true.
+          case("directories")
+            directories_present = .true.
+          case("macros")
+            macros_present = .true.
+         end select
+      end do
+   end subroutine check
+
+   !> Construct new preprocess array from a TOML data structure.
+   subroutine new_preprocessors(preprocessors, table, error)
+
+      !> Instance of the preprocess configuration
+      type(preprocess_config_t), allocatable, intent(out) :: preprocessors(:)
+
+      !> Instance of the TOML data structure
+      type(toml_table), intent(inout) :: table
+
+      !> Error handling
+      type(error_t), allocatable, intent(out) :: error
+
+      type(toml_table), pointer :: node
+      type(toml_key), allocatable :: list(:)
+      integer :: iprep, stat
+
+      call table%get_keys(list)
+
+      ! An empty table is not allowed
+      if (size(list) == 0) then
+         call syntax_error(error, "No preprocessors defined")
+      end if
```

Comment:
UI question for @arteevraina and @fortran-lang/fpm: Should just

```toml
[preprocess]
```

be allowed and imply

```toml
[preprocess]
[preprocess.cpp]
```
?

I'm split. On one side, explicit is more clear. On the other, requiring `[preprocess.cpp]` adds the needed boilerplate if one just wants CPP preprocessing everywhere as a sane default, it could be nice to just type `[preprocess]` in the manifest and be done with it, and a sane default philosophy is I think aligned with the original vision for fpm. But it would effectively make CPP the first-class preprocessor, for better or for worse.

Reply:
I would say being explicit about which preprocessors we want to use while building fortran projects using fpm would make the `manifest` file clear and what goes under the hood. But, if we are going to make `cpp` as first choice preprocessors for fpm-based projects then I think simply `[preprocess]` should be fine. I think more or less this decision depends on how we expect to see fpm in the future. Either it should have a favorite `cpp` preprocessor with allowing other preprocessors as well or just being preprocessor agnostic and leaving it to the user to be explicit.

Reply:
I would prefer to be explicit too. For example, if a compiler does not support `cpp`, the default option would fail. Furthermore we don't know if `cpp` will still be the "default"/preferred preprocessor in the future.